odoo/tools/populate.py

```diff
@@ -59,12 +59,26 @@ def get_field_variation_date(model: Model, field: Field, factor: int, series_ali
     setting duplicated records too far back in the past.
     """
     total_days = min((MAX_DATETIME - MIN_DATETIME).days, factor)
-    return SQL("""
-        %(field)s - (%(factor)s - %(series_alias)s) * floor(%(total_days)s/%(factor)s) * interval '1 days'
-    """, field=SQL.identifier(field.name),
+    cast_type = SQL('date') if field.type == 'date' else SQL('timestamp')
```

Comment:
you can use `field._column_type[1]` which is `'date'` for Date and `'timestamp'` for Datetime